include/envoy/buffer/buffer.h

```diff
@@ -194,6 +197,118 @@ class Instance {
    * -1 for failure. If the call is successful, errno_ shouldn't be used.
    */
   virtual Api::SysCallIntResult write(int fd) PURE;
+
+  /**
+   * Copy an integer out of the buffer.
+   * @param start supplies the buffer index to start copying from.
+   * @param Size how many bytes to read out of the buffer.
```

Comment:
Ok. I still think it needs at least a brief comment.

Does the `peekBEInt` implementation need to be modified? As I read it, invoking `peekBEInt<uint32_t, 3>()` on a buffer containing (hex bytes) `01 02 03 04` will return `0x020304`. If the goal is to support uncommon integer sizes, I think it should return `0x010203`. 

Also, if you call `peekBEInt<uint32_t, 3>()` on a buffer containing only 3 bytes, the length check will succeed, but it will try to read past the end of the buffer.

Reply:
I also wonder if this code should handle signed integers differently? If I invoke `peekLEInt<int32_t, 3>()` on a buffer containing `FF FF FF` do I get -1 or 0xFFFFFF in my int32_t?

Reply:
On a second glance at the code, I don't think `01 02 03 04` `peekBEInt<uint32_t, 3>(start = 0)` will read `02 03 04`, though. Here's different passes at the code, with proper variable substitution:
```
     static_assert(Size <= sizeof(T), "requested size is bigger than integer being read");

     if (length() < start + Size) {
      throw EnvoyException("buffer underflow");
    }
     constexpr const auto displacement = Endianness == ByteOrder::BigEndian ? sizeof(T) - Size : 0;
     auto result = static_cast<T>(0);
    copyOut(start, Size, reinterpret_cast<char*>(std::addressof(result)) + displacement);
    return fromEndianness<Endianness>(result);
```
becomes
```
    static_assert(3 <= 4, "requested size is bigger than integer being read");

    if (4 < 0 + 3) {
      throw EnvoyException("buffer underflow");
    }
     constexpr const auto displacement = true ? 4 - 3 : 0;
     uint32_t result = 0;
    copyOut(0, 3, reinterpret_cast<char*>(std::addressof(result)) + displacement);
    return fromEndianness<BigEndian>(result);
```
which becomes
```
     constexpr const auto displacement = 1;
     uint32_t result = 0;
    copyOut(0, 3, reinterpret_cast<char*>(std::addressof(result)) + 1);
    return fromEndianness<BigEndian>(result);
```
which copies the first 3 bytes from the buffer, namely `01 02 03` into the last three bytes of `result`, yielding `00 01 02 03`. Then `fromEndianness` properly converts that into host byte order.

Nevertheless, the expected result you listed is indeed the correct expected behavior. As long as we add some tests and make sure they pass, we should be safe.

Reply:
We still need to implement sign extension, though. Let me sketch something quick for it...

Reply:
I believe we could replace
```
    copyOut(start, Size, reinterpret_cast<char*>(std::addressof(result)) + displacement);
    return fromEndianness<Endianness>(result);
```
with
```
    auto bytes = reinterpret_cast<char*>(std::addressof(result));
    copyOut(start, Size, &bytes[displacement]);

    constexpr const auto most_significant_read_byte = Endianness == ByteOrder::BigEndian
      ? displacement
      : Size - 1;

    constexpr const auto all_bits_enabled = static_cast<T>(~static_cast<T>(0));
    constexpr const auto extension = std::is_signed<T>::value && Size < sizeof(T) && bytes[most_significant_read_byte] < 0
      // integer overflow-free sign extension, assuming two's complement
      ? Endianness == ByteOrder::BigEndian
        ? static_cast<T>(all_bits_enabled << (Size * CHAR_BIT))
        : static_cast<T>(all_bits_enabled >> (Size * CHAR_BIT))
      // no sign extension necessary
      : static_cast<T>(0);
    return fromEndianness<Endianness>(static_cast<T>(result | extension));
```

Reply:
updated comment snippet reflecting sign extension:

--- BEGIN COMMENT SNIPPET ---
Some protocols have integer fields whose size in bytes won't match the size in bytes of C++'s integer types.

Take a 3-byte integer field for example, which we want to represent as a 32-bit (4 bytes) integer. One option to deal with that situation is to read 4 bytes from the buffer and ignore 1. There are a few problems with that solution, though.

The first problem is buffer underflow: there may not be more than `Size` bytes available (say, last field in the payload), so that's an edge case to take into consideration.

The second problem is draining the buffer after reading. With the above solution we cannot read and discard in one go. We'd need to `peek` 4 bytes, ignore 1 and then `drain` 3. That not only looks hacky since the sizes don't match, but also produces less terse code and requires the caller to propagate that logic to all call sites.

Things complicate even further when endianness is taken into consideration: should the most or least-significant bytes be padded? Dealing with this situation requires a high level of care and attention to detail. Properly calculating which bytes to discard and how to displace the data is not only error prone, but also shifts to the caller a burden that could be solved in a much more generic, transparent and well tested manner.

The last problem in the list is sign extension, which should be properly handled when reading signed types with negative values.

To make matters easier, the optional `Size` parameter can be specified in those situations where there's a need to read less bytes than a C++'s integer size in bytes.

For the most common case when one needs to read exactly as many bytes as the size of C++'s integer, this parameter can simply be omitted and it will be automatically deduced from the size of the type `T`.
---  END COMMENT SNIPPET  ---

Reply:
Working on this, thanks @juchem !

Reply:
Sorry, yes I misread how the displacement was used. It does take the correct bytes for big-endian peek.

For sign extension, I don't think you need the ternary expression. You want to `all_bits_enabled << (Size * CHAR_BIT)`. Otherwise for little endian you'll be modifying the low-order bits.

Also, perhaps `static_cast<T>(~static_cast<T>(0))` can be reduced to just `~static_cast<T>(0)`?

Reply:
I had problems with type promotion and the complement operator in the past. Honestly, I can never remember whether it's needed before or after `~` so I just choose to err on the safe side.

With regards to the ternary, I'm not so sure, but you may be right.

My thinking was this (3 bytes read into a 32-bit signed integer from buffer with `f1 f2 f3 f4`):

Big Endian:
- `result` is initialized with `0`;
- bytes are read at displacement `1`, `result = 00 f1 f2 f3`;
- we need to fill the leftmost bytes due to big-endianness (most significant byte is to the left), therefore `extension` should be `ff 00 00 00`, so shift `all_bits_enabled` left by `Size - 1` bytes.

Little Endian:
- `result` is initialized with `0`;
- bytes are read at displacement `0`, `result = f1 f2 f3 00`;
- we need to fill the righttmost bytes due to little-endianness (most significant byte is to the right), therefore `extension` should be `00 00 00 ff`, so shift `all_bits_enabled` right by `Size - 1` bytes.

The tests should surface that anyway, so I'm not too concerned.